### PR TITLE
Add support for multiple HTTP connection backends and Apache's HttpClient backend support 

### PR DESCRIPTION
**Pull request by dgiagio**
- Add initial infrastructure for multiple HTTP connection backends.
- Add support for Apache HttpClient as HTTP connection backend.
- Apache HttpClient backend supports proxy, authenticated or not.

### CODE REVIEW

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackHttpClientConnectionFactory.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Implementation of a factory for connections using Apache's HttpClient.
+ *
+ * @see com.suse.saltstack.netapi.client.SaltStackHttpClientConnection
+ */
+public class SaltStackHttpClientConnectionFactory implements SaltStackConnectionFactory {
+    @Override
+    public SaltStackHttpClientConnection create(String endpoint,
+                                         SaltStackClientConfig config) {
```
Comment (renner) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackHttpClientConnectionFactory.java]: Please do not use such block indentation, rather indent 8 spaces after breaking a long line (as [mentioned] (https://github.com/SUSE/saltstack-netapi-client-java/pull/1#issuecomment-72463963)).

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackJDKConnectionFactory.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Implementation of a factory for connections using JDK's HttpURLConnection.
+ *
+ * @see SaltStackJDKConnection
+ */
+public class SaltStackJDKConnectionFactory implements SaltStackConnectionFactory {
+    @Override
+    public SaltStackJDKConnection create(String endpoint,
+                                         SaltStackClientConfig config) {
```
Comment (renner) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackJDKConnectionFactory.java]: Do not use block indentation, please indent 8 spaces.

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackHttpClientConnection.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    private final SaltStackClientConfig config;
+
+    /**
+     * Init a connection to a given SaltStack API endpoint.
+     *
+     * @param endpointIn the endpoint
+     * @param configIn the config
+     */
+    public SaltStackHttpClientConnection(String endpointIn,
+                                         SaltStackClientConfig configIn) {
```
Comment (renner) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackHttpClientConnection.java]: Do not use block indentation, please indent 8 spaces.

**Comment by dgiagio** [comment]
@renner Fixed indent. Thanks for the heads up!

**Comment by dgiagio** [comment]
@moio Javadoc fixed as suggested.

**Comment by renner** [comment]
There is two jars you added to the libs that seem to be optional for what we are doing, could you please remove these?

- httpclient-cache-4.3.6.jar
- httpmime-4.3.6.jar

**Comment by dgiagio-druid** [comment]
@renner Removed, thanks.

**Comment by dgiagio-druid** [comment]
@renner ```ant run-tests``` works OK after unnecessary jars removal.

**Comment by renner** [comment]
Ok, merged it. Please note that I added some more javadoc and annotations that were still missing as mentioned before by @moio, see ecf5f78868ed41f878598d965a9e48cf49e9eed4. I further made the *HttpComponents* based backend the default with 2603c9b0318eabfec126173bc7e1272fe12aed59.

Thanks for your contribution!

**Comment by dgiagio** [comment]
Great, thanks!

**Pull request merged**